### PR TITLE
Weekly patterns discontinuity provider

### PR DESCRIPTION
**Pull request by murcikan-scottlogic**

### CODE REVIEW

File: packages/d3fc-discontinuous-scale/src/discontinuity/skipWeeklyPattern.js [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    /**
+     * When given a value falls within a discontinuity (i.e. an excluded domain range) it should be shifted forwards to the discontinuity boundary. 
+     * Otherwise, it should be returns unchanged.
+     * @param {Date} date - date to clamp up
+     * @returns {Date}
+     */
+    instance.clampUp = (date) => {
+        const tradingDay = tradingDays[dateTimeUtility.getDay(date)];
+
+        for (let i = 0; i < tradingDay.nonTradingTimeRanges.length; i++) {
```
Comment (ColinEberhardt) [inline packages/d3fc-discontinuous-scale/src/discontinuity/skipWeeklyPattern.js]: Perhaps consider using [for ... of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of#syntax)?

File: packages/d3fc-discontinuous-scale/src/discontinuity/skipWeeklyPattern.js [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+    /**
+     * When given a pair of values, this function returns the distance between the, in domain units, minus any discontinuities. discontinuities.
+     * @param {Date} startDate 
+     * @param {Date} endDate 
+     * @returns {number} - the number of milliseconds between the dates
+     */
+    instance.distance = (startDate, endDate) => {
+
+        const sortChronologically = (start, end) => {
```
Comment (ColinEberhardt) [inline packages/d3fc-discontinuous-scale/src/discontinuity/skipWeeklyPattern.js]: This function is only used once, perhaps inline?

File: packages/d3fc-discontinuous-scale/src/discontinuity/skipWeeklyPattern/tradingDay.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
+import { nonTradingTimeRange } from './nonTradingTimeRange';
+import { dayBoundary, millisPerDay } from './constants';
+
+/**
+ * Represents a Trading day
+ * @param { string[][] } rawDicontinuityTimeRanges - Array of time range tuples e.g. [["07:45", "08:30"), ["19:00:45.500", "EOD")]
```
Comment (ColinEberhardt) [inline packages/d3fc-discontinuous-scale/src/discontinuity/skipWeeklyPattern/tradingDay.js]: `Dicontinuity` sp.

File: scripts/jest/jest.config.js [hunk 4]
```diff
@@ -1,6 +1,8 @@
+process.env.TZ = "Europe/London";
 module.exports = {
     rootDir: '../../',
     roots: ['<rootDir>/packages'],
     setupFilesAfterEnv: [require.resolve('./setup.js')],
+    moduleNameMapper: { '^d3-(.*)$': `d3-$1/dist/d3-$1` },
```
Comment (ColinEberhardt) [inline scripts/jest/jest.config.js:6]: just out of interest - what does this do?

**Review by ColinEberhardt — COMMENTED** [review]
This looks really good - thanks for the comprehensive tests. I've added a few very minor comments.

One thought I did have, I can see that the implementation does a bit of string arsing / formatting. Does that have much of a performance impact? When rendering a chart, the `offset` and `distance` functions are called quite often.

I'd be happy to merge if you can also update the README to briefly document this discontinuity provider

File: scripts/jest/jest.config.js (hunk 4 above)
Comment (murcikan-scottlogic) [reply scripts/jest/jest.config.js:6]: It's to solve the following issue: https://github.com/facebook/jest/issues/12036

File: packages/d3fc-discontinuous-scale/src/discontinuity/skipWeeklyPattern.js (hunk 1 above)
Comment (murcikan-scottlogic) [reply packages/d3fc-discontinuous-scale/src/discontinuity/skipWeeklyPattern.js]: done

File: packages/d3fc-discontinuous-scale/src/discontinuity/skipWeeklyPattern.js (hunk 2 above)
Comment (murcikan-scottlogic) [reply packages/d3fc-discontinuous-scale/src/discontinuity/skipWeeklyPattern.js]: done

File: packages/d3fc-discontinuous-scale/src/discontinuity/skipWeeklyPattern/tradingDay.js (hunk 3 above)
Comment (murcikan-scottlogic) [reply packages/d3fc-discontinuous-scale/src/discontinuity/skipWeeklyPattern/tradingDay.js]: corrected

**Review by ColinEberhardt — APPROVED** [review]

**Pull request merged**